### ISSUE TITLE
TestNG swallowing an IllegalArgumentException when DataProivder mismatches Test

### ISSUE THREAD

**Issue by dklaus** [cbeust/testng#123]
I have hit an issue where TestNG will hide a `IllegalArgumentException` when it throws one and the exception is marked as an `expectedExceptions`.  This happens when TestNG is matching a dataProvider up to a test's parameters, the data does not match up, and `IllegalArgumentException` is an `expectedExceptions`.  An example of this is:

``` java
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

public class swallowedException
{
    @DataProvider
    public Object[][] testInts()
    {
        return new Object[][] {
                new Object[] {new Integer(4) },
                new Object[] {new Integer(8) },
                new Object[] {new Integer(12) } };
    }

    @Test(dataProvider = "testInts", expectedExceptions = IllegalArgumentException.class)
    public void theTest(String aString)
    {
        System.out.println("An item is \"" + aString + "\"");
    }
}
```

In this case TestNG will throw an `IllegalArgumentException` when trying to match up the Integer from the dataProvider to the String parameter in the test (which is as expected.)   However, because `IllegalArgumentException` is marked as an expected exception in the test, TestNG is marking this test as `PASSED` (all 3 times) when in fact it was not run.  For this example, if you remove the `expectedExceptions` annotation, TestNG marks the test as `FAILED` with the exception thrown by TestNG.  The actual code being tested was expected to throw an `IllegalArgumentException`, was not being tested, and was marked by TestNG as PASSED.

**Comment by dimo414** [comment]
Just ran into this myself, definitely unexpected behavior.

**Comment by juherr** [comment]
@dimo414 Are you open to try to fix it?

**Comment by dimo414** [comment]
@juherr I can't promise anything, but I'll certainly send a pull request if I have time.

**Comment by krmahadevan** [comment]
@dimo414 
This problem is no longer recreatable using **TestNG 6.10**.

When you run the same sample here's what I see:

```
org.testng.internal.reflect.MethodMatcherException: 
Data provider mismatch
Method: theTest([Parameter{index=0, type=java.lang.String, declaredAnnotations=[]}])
Arguments: [(java.lang.Integer)4]

	at org.testng.internal.reflect.DataProviderMethodMatcher.getConformingArguments(DataProviderMethodMatcher.java:52)
	at org.testng.internal.Invoker.injectParameters(Invoker.java:1278)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1171)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.TestRunner.privateRun(TestRunner.java:756)
	at org.testng.TestRunner.run(TestRunner.java:610)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:387)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:382)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1293)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1218)
	at org.testng.TestNG.runSuites(TestNG.java:1133)
	at org.testng.TestNG.run(TestNG.java:1104)
	at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:72)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:124)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
```

@juherr - Do you think we can go ahead and close this issue ?

**Comment by juherr** [comment]
@krmahadevan Sure, but could you check if the case is covered by a test? It will avoid potential regression

**Comment by dimo414** [comment]
Looks like this is fixed, thanks! I think f3e44d60bf67 was the fix (since it added the "Data provider mismatch" message in `DataProviderMethodMatcher`). This should probably be called out in the release notes (if it's not already) since it'll cause some "passing" tests to start failing.
